### PR TITLE
[llk_helper_library] 5/8 eltwise helpers: migrate normalization kernels

### PR DESCRIPTION
**Pull request by astancovTT**
# [llk_helper_library] 5/8 eltwise helpers: migrate normalization kernels

## Summary

This is PR 5 of the #49991 split. It is stacked on PR 4 and, through that
parent, the PR 1 helper framework. It migrates normalization compute kernels to
`eltwise_chain` while preserving their existing TTNN APIs and behavior.

Each migration replaces handwritten DFB lifecycle code, operation
initialization, DEST management, and tile loops with explicit chain elements.
`compute_kernel_hw_startup(...)` remains caller-owned.

## Migrated kernels

- BatchNorm and running-statistics kernels, including SFPU variants.
- GroupNorm regular, sharded-v2, Welford, and Welford sharded-v2 kernels.
- LayerNorm regular, large-tensor, and Welford variants.
- Distributed LayerNorm/RMSNorm pre- and post-all-gather kernels.
- Attention softmax and sharded attention softmax kernels.

## Migration details

- The existing math-fidelity, FP32 DEST, unpack-mode, and data-format
  reconfiguration behavior stays explicit at the migration boundaries.
- GroupNorm and LayerNorm retain their staged reduction, affine gamma/beta,
  tilize/untilize, and output paths rather than combining independently
  initialized stages.
- Distributed normalization retains its all-gather staging and removes the
  legacy local chain helper after moving consumers to the shared framework.
- The planned Welford follow-up restores the FP32 SrcA format before each
  batch's transpose initialization. This is necessary because the migrated
  tail stage changes SrcA; without the restore, a later FP32 batch can run with
  stale unpack configuration.
- The SDXL GroupNorm performance threshold remains aligned with the migrated
  kernel's measured baseline.

## Review guide

Suggested review order:

1. BatchNorm/running-statistics kernels for the basic normalization chains.
2. GroupNorm and Welford GroupNorm variants, especially FP32 transpose and
   per-batch reconfiguration.
3. LayerNorm and distributed LayerNorm/RMSNorm for their staged reductions and
   communication boundaries.
4. Softmax kernels for attention-specific masking and DEST accumulation.

## Scope

This PR includes two targeted changes to the shared helper framework from PR 1:
Cumulative-wait inputs own their CB window regardless of pop policy, preventing
conflicting peer pops; and a debug assertion rejects FullBlock tail synchronization
with upfront reservation and per-block pushes. Both have regression-test coverage.

The eltwise/generator, copy/data-movement, Moreh, Moreh-training, and TT-Train
migrations remain in their respective split PRs. The experimental migrations
come from the parent PR.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:astancov/eltwise_helper_migration_5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:astancov/eltwise_helper_migration_5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:astancov/eltwise_helper_migration_5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:astancov/eltwise_helper_migration_5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:astancov/eltwise_helper_migration_5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:astancov/eltwise_helper_migration_5)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax_sharded.cpp [hunk 1]
```diff
@@ -72,163 +58,73 @@ void kernel_main() {
...
+            ckl::Optional<!numeric_stable, ckl::Exp<static_cast<ckl::Approx>(EXP_APPROX), ckl::Dst::D0>>{},
+            ckl::PackTile<ckl::output(
+                dfb_x_id, ckl::ReservePolicy::Upfront, ckl::PushPolicy::AtEnd, ckl::DataFormatReconfig::Disabled)>{});
 
 // add numeric_stable
 // fuse exp with sub tiles
 #ifdef NUMERIC_STABLE
         dfb_x_obj.wait_front(block_w);
-        calc_numeric_stable<block_w, num_subblocks_w, subblock_w, dfb_x, dfb_max_scaler, dfb_max, dfb_exps>();
+        calc_numeric_stable<block_w, num_subblocks_w, subblock_w, dfb_x_id, dfb::max_scaler, dfb::max, dfb::exps>();
```
Comment (adamjanovicTT) [inline ttnn/cpp/ttnn/operations/normalization/softmax/device/kernels/attention/compute/softmax_sharded.cpp]: `num_subblocks_w` is no longer used in the latest `calc_numeric_stable` implemenation.

**Review by mbezuljTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm_large_tensor_welford.cpp [hunk 2]
```diff
@@ -495,40 +484,22 @@ void kernel_main() {
...
-        tile_regs_commit();
-        dfb_ex2_obj.pop_front(onetile);
-
-        dfb_ex2pe_obj.reserve_back(onetile);
-        tile_regs_wait();
-        pack_tile(dst0, dfb_ex2pe);
-        tile_regs_release();
-        dfb_ex2pe_obj.push_back(onetile);
+        ckl::eltwise_chain(
+            ckl::IterationShape::tiles(onetile),
```
Comment (adamjanovicTT) [inline ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm_large_tensor_welford.cpp]: We should probably use the shorter version: `ckl::IterationShape::one_tile()`?

File: ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm_large_tensor_welford.cpp [hunk 3]
```diff
@@ -495,40 +484,22 @@ void kernel_main() {
...
-        dfb_ex2pe_obj.pop_front(onetile);
-        tile_regs_commit();
-
-        dfb_ex2pe_obj.reserve_back(onetile);
-        tile_regs_wait();
-        pack_tile(dst0, dfb_ex2pe);
-        tile_regs_release();
-        dfb_ex2pe_obj.push_back(onetile);
+        ckl::unary_bcast<ckl::BroadcastDim::Col, ckl::input(dfb_ex2pe), ckl::output(dfb_ex2pe)>(
+            ckl::IterationShape::tiles(onetile));
```
Comment (adamjanovicTT) [inline ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm_large_tensor_welford.cpp]: Fix here too and other places if they exist: [original comment](https://github.com/tenstorrent/tt-metal/pull/55698/changes#r4062762365)

**Review by bbradelTT — APPROVED** [review]
Please comment on or address AI comments.

**Review by adamjanovicTT — APPROVED** [review]
